airflow/plugins/operators/scrape_state_geoportal.py

```diff
@@ -0,0 +1,224 @@
+import gzip
+import logging
+import os
+from typing import ClassVar, List
+
+import pandas as pd  # type: ignore
+import pendulum
+import requests
+from calitp_data_infra.storage import PartitionedGCSArtifact, get_fs  # type: ignore
+from pydantic import HttpUrl, parse_obj_as
+
+from airflow.models import BaseOperator  # type: ignore
+
+API_BUCKET = os.environ["CALITP_BUCKET__STATE_GEOPORTAL_DATA_PRODUCTS"]
+
+
+class StateGeoportalAPIExtract(PartitionedGCSArtifact):
+    bucket: ClassVar[str]
+    execution_ts: pendulum.DateTime = pendulum.now()
+    dt: pendulum.Date = execution_ts.date()
+    partition_names: ClassVar[List[str]] = ["dt", "execution_ts"]
+
+    # The name to be used in the data warehouse to refer to the data
+    # product.
+    product: str
+
+    # The root of the ArcGIS services. As of Nov 2024, this should
+    # be "https://caltrans-gis.dot.ca.gov/arcgis/rest/services/".
+    root_url: str
+
+    # The name of the service being requested. In the feature service's
+    # URL, this will be everything between the root and "/FeatureServer".
+    # Don't include a leading or trailing slash.
+    service: str
+
+    # The layer to query. This will usually be "0", so that is the
+    # default.
+    layer: str = "0"
+
+    # The query filter. By default, all rows will be returned from the
+    # service. Refer to the ArcGIS documentation for syntax:
+    # https://developers.arcgis.com/rest/services-reference/enterprise/query-feature-service-layer/#request-parameters
+    where: str = "1=1"
+
+    # A comma-separated list of fields to include in the results. Use
+    # "*" (the default) to include all fields.
+    outFields: str = "*"
+
+    # The number of records to request for each API call (the operator
+    # will request all data from the layer in batches of this size).
+    resultRecordCount: int
+
+    @property
+    def table(self) -> str:
+        return self.product
+
+    @property
+    def filename(self) -> str:
+        return self.table
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def fetch_from_state_geoportal(self):
+        """ """
+
+        logging.info(f"Downloading state geoportal data for {self.product}.")
+
+        try:
+            # Set up the parameters for the request
+            url = f"{self.root_url}/{self.service}/FeatureServer/{self.layer}/query"
+            validated_url = parse_obj_as(HttpUrl, url)
+
+            params = {
+                "where": self.where,
+                "outFields": self.outFields,
+                "f": "geojson",
+                "resultRecordCount": self.resultRecordCount,
+            }
+
+            all_features = []  # To store all retrieved rows
+            offset = 0
+
+            while True:
+                # Update the resultOffset for each request
+                params["resultOffset"] = offset
+
+                # Make the request
+                response = requests.get(validated_url, params=params)
```

Comment:
**question:** What should happen if we get a non-200 response from the server? I would think we should raise an exception so that the task fails and has to retry. Does that seem right? You could add `.raise_for_status()` at the end of the line if so. See https://3.python-requests.org/user/quickstart/#response-status-codes

Reply:
added! great suggestion